app/intune/index.js

```diff
@@ -0,0 +1,91 @@
+const dbus = require('@homebridge/dbus-native');
+const { LucidLog } = require('lucid-log');
+
+var sessionBus = dbus.sessionBus();
+
+var intuneAccount = null;
+
+var brokerService = sessionBus.getService('com.microsoft.identity.broker1');
+
+function processInTuneAccounts(logger, resp, ssoInTuneAuthUser) {
+	response = JSON.parse(resp);
+	if ('error' in response) {
+		logger.warn('Failed to retrieve InTune account list: ' + response.error.context);
+		return;
+	};
+
+	if (ssoInTuneAuthUser == '') {
+		intuneAccount = response.accounts[0];
+		logger.debug('Using first available InTune account (' + intuneAccount.username + ')');
+	} else {
+		for (account in response.accounts) {
+			if (account.username == ssoIntuneAuthUser) {
+				intuneAccount = account;
+				logger.debug('Found matching InTune account (' + intuneAccount.username + ')');
+				break;
+			}
+		}
+		if (intuneAccount == null) {
+			logger.warn('Failed to find matching InTune account for ' + ssoIntuneAuthUser + '.');
+		}
+	}
+}
+
+exports.initSso = function initIntuneSso(logger, ssoInTuneAuthUser) {
+	logger.debug("Initializing InTune SSO");
+	brokerService.getInterface(
+		'/com/microsoft/identity/broker1',
+		'com.microsoft.identity.Broker1', function(err, broker) {
+			if (err) {
+				logger.warn('Failed to find microsoft-identity-broker DBus interface');
+				return;
+			}
+			broker.getAccounts('0.0', '', JSON.stringify({'clientId': '88200948-af09-45a1-9c03-53cdcc75c183', 'redirectUri':'urn:ietf:oob'}), function(err, resp) {
+				if (err) {
+					logger.warn('Failed to communicate with microsoft-identity-broker');
+					return;
+				}
+				processInTuneAccounts(logger, resp, ssoInTuneAuthUser);
+			});
+		});
+}
+
+exports.setupUrlFilter = function setupUrlFilter(filter) {
+	filter.urls.push('https://login.microsoftonline.com/*');
+}
+
+exports.isSsoUrl = function isSsoUrl(url) {
+	return intuneAccount != null && url.startsWith('https://login.microsoftonline.com/');
+}
+
+function processPrtResponse(logger, resp, detail) {
+	response = JSON.parse(resp);
+	if ('error' in response) {
+		logger.warn('Failed to retrieve Intune SSO cookie: ' + response.error.context);
+	} else {
+		logger.debug('Adding SSO credential');
+		detail.requestHeaders['X-Ms-Refreshtokencredential'] = response['cookieContent'];
+	}
+}
+
+exports.addSsoCookie = function addIntuneSsoCookie(logger, detail, callback) {
+	logger.debug('Retrieving InTune SSO cookie');
+	if (intuneAccount == null) {
+		logger.info("InTune SSO not active");
+		callback({
+			requestHeaders: detail.requestHeaders
+		});
+		return;
+	}
+	brokerService.getInterface(
+		'/com/microsoft/identity/broker1',
+		'com.microsoft.identity.Broker1', function(err, broker) {
+			broker.acquirePrtSsoCookie('0.0', '', JSON.stringify({'ssoUrl':detail.url, 'account':intuneAccount, 'authParameters':{'authority':'https://login.microsoftonline.com/common/'}}), function(err, resp) {
```

Comment:
@KrissN Last week I developed a plugin for Firefox on Linux that uses exactly the same approach. This will be OSSed soon. I'm currently working on a similar thing for EWS in Evolution.

I'm just wondering if the `authParameters` part here is really sufficient (apparently it works), as the Edge Browser puts more data in there:

Sample request from Edge

```
            'accessTokenToRenew': '',
            'account': account,
            'additionalQueryParametersForAuthorization': {},
            'authority': 'https://login.microsoftonline.com/common',
            'authorizationType': 8,  # OAUTH2
            'clientId': EDGE_BROWSER_CLIENT_ID,
            'decodedClaims': '',
            'enrollmentId': '',
            'password': '',
            'popParams': None,
            'redirectUri': 'https://login.microsoftonline.com'
                           '/common/oauth2/nativeclient',
            'requestedScopes': scopes,
            'username': account['username'],
            'uxContextHandle': -1
```

The interesting parts here are:

- repetition of the account
- authorization type
- client ID
- requested scopes (which usually is just `https://graph.microsoft.com/.default`)
- username

Did you find documentation about the exact interface, or also just reverse engineered it?

Reply:
Also here, a client ID can be passed. I recommend to use the one from edge here, as Edge itself is technically more or less a chrome (webkit). An alternative would be to make it configurable.

Reply:
No, I have found no documentation about this DBus interface. I have reverse-engineered it, in a way that I have found a minimum set of arguments that have worked.

Reply:
@KrissN thanks for the confirmation. It would be good if we could at least make the client id configurable. Apart from that, we should also ensure that the service is running before calling it on DBus. Otherwise, the sign-in does not work directly after boot (before another application like Edge activates it via DBus). For reference, you might want to have a look how I implemented that in the Firefox / Chrome plugin: https://github.com/siemens/linux-entra-sso/blob/main/linux-entra-sso.py#L63